otherlibs/stdune-unstable/temp.ml

```diff
@@ -94,7 +94,8 @@ let clear_dir dir =
 
 let temp_path =
   try_paths 1000 ~f:(fun candidate ->
-      if Path.exists candidate then raise Exit;
+      if Path.exists candidate then
+        raise (Failure "the candidate temp path has already been taken");
```

Comment:
We don't use `Failure` in stdune. By the way, `try_paths` seems wrong to me. It should only be catching `Exit` errors and retrying on those rather than on arbitrary failures.

Reply:
How would you suggest to give a meaningful message in the (unlikely) scenario that we can't generate an untaken random path?

Reply:
Such a message might show up somewhere in the log along a "cache store error", and I thought that seeing `Exit` there might not be very informative.

Reply:
`User_error.raise` is what we usually use for a "nice message". `Code_error.raise` is what we use for programming errors in dune. One of these should be good.

Reply:
> How would you suggest to give a meaningful message in the (unlikely) scenario that we can't generate an untaken random path?

Thinking about this, if this unlikely scenario ever happened we should just change our code to read the directory and find a filename that does not exist. So if we fail to generate a valid filename, we should just do a `Code_error.raise` since it's a bug in our code that we need to fix.

Reply:
I agree that using `Code_error` here sounds right, but then I don't see how to make that compatible with your other point about `Exit`. I guess one way out is to introduce a special exception `Retry` or something like that, and then raise the `Code_error` after all tries fail. Does that sound right?

Reply:
> I guess one way out is to introduce a special exception Retry or something like that, and then raise the Code_error after all tries fail. Does that sound right?

Exactly. Note that `Exit` is basically used as `Retry` here, but making a dedicated exception would be cleaner. Not sure if you're familiar with the convention, but `Exit` in OCaml is explicitly reserved for control flow and caller code should never see it. I just got a little lazy here.

Reply:
Interesting, I didn't know about that convention. OK, I'll put up a PR with `Retry` and `Code_error` as discussed. Thanks!